### ISSUE TITLE
Associate Event and HcbCode

### ISSUE THREAD

**Issue by garyhtou** [hackclub/hcb#3426]
Not sure if @albertchae was already planning to do this, but it'd be very handy to be able to do:
```ruby
Event.find(1).hcb_codes
```

and vice versa. Currently `HcbCode.find(1).events` is implemented, but it isn't an association

---

Here's what I have so far, but it seems very messy

## Event -> HcbCode

```ruby
# event.rb

has_many :hcb_codes, ->(event) {
  unscope(where: :event_id)
    .left_joins(:canonical_transactions, :canonical_pending_transactions)
    .left_joins(canonical_transactions: :canonical_event_mapping, canonical_pending_transactions: :canonical_pending_event_mapping)
    .where('canonical_event_mappings.event_id = ? OR canonical_pending_event_mappings.event_id = ?', event.id, event.id)
}


# based on https://stackoverflow.com/questions/36749740/activerecord-has-many-through-through-multiple-models
```

## HcbCode -> Event

```ruby
# hcb_code.rb

has_many :events, ->(hcb_code) {
  unscope(where: :hcb_code_id)
    .left_joins(:canonical_transactions, :canonical_pending_transactions)
    .where("canonical_transactions.hcb_code = ? OR canonical_pending_transactions.hcb_code = ?", hcb_code.hcb_code, hcb_code.hcb_code)
}
```
This is incorrect since it seems like HcbCode relies on more than just CanonicalTransaction and CanonicalPendingTransaction to get the event.
https://github.com/hackclub/bank/blob/94a91292ec187a16d0f125c55f1e837cdf549144/app/models/hcb_code.rb#L117-L148

https://heroku-app97991095.airbrake.io/projects/288439/groups/3449473709812445527?tab=overview

**Comment by albertchae** [comment]
Hey @garyhtou , this is definitely on my list of associations to untangle, but it's a pretty gnarly one. I have some ideas on how to approach it and I saw you added some telemetry to see if `try`ing on some of the associations makes sense. Do you want to pair on it sometime?

**Comment by garyhtou** [comment]
Yes! @albertchae, I'd love to. let's find a time sometime next week or later

**Comment by albertchae** [comment]
@garyhtou cool I'll ping you on slack to figure out a time

**Comment by albertchae** [comment]
🥲 
<img width="370" height="50" alt="Image" src="https://github.com/user-attachments/assets/f287953b-8f8f-4adb-a923-275f91352e40" />
